src/spatial/directcrrssa.jl

```diff
@@ -0,0 +1,265 @@
+    rt = PriorityTable(ratetogroup, zeros(T, 1), minrate, 2 * minrate)
+
+    DirectCRRSSAJumpAggregation{
+        T,
+        BD,
+        M,
+        RNG,
+        J,
+        RX,
+        HOP,
+        typeof(dg),
+        typeof(vtoj_map),
+        typeof(jtov_map),
+        SS,
+        typeof(rt),
+        Nothing,
+        Nothing,
+        Nothing,
+    }(nj, nj, njt, et, bd, u_low_high, rx_rates, hop_rates, site_rates_high, sps, rng, dg,
+        vtoj_map, jtov_map, spatial_system, num_specs, rt, nothing, nothing)
+end
+
+############################# Required Functions ##############################
+# creating the JumpAggregation structure (function wrapper-based constant jumps)
+function aggregate(aggregator::DirectCRRSSA, starting_state, p, t, end_time,
+    constant_jumps, ma_jumps, save_positions, rng; hopping_constants,
+    spatial_system, bracket_data = nothing, kwargs...)
+    T = typeof(end_time)
+    num_species = size(starting_state, 1)
+    majumps = ma_jumps
+    if majumps === nothing
+        majumps = MassActionJump(Vector{T}(),
+            Vector{Vector{Pair{Int, Int}}}(),
+            Vector{Vector{Pair{Int, Int}}}())
+    end
+
+    next_jump = SpatialJump{Int}(typemax(Int), typemax(Int), typemax(Int)) #a placeholder
+    next_jump_time = typemax(T)
+    rx_rates = LowHigh(RxRates(num_sites(spatial_system), majumps),
+        RxRates(num_sites(spatial_system), majumps);
+        do_copy = false) # do not copy ma_jumps
+    hop_rates = LowHigh(HopRates(hopping_constants, spatial_system),
+        HopRates(hopping_constants, spatial_system);
+        do_copy = false) # do not copy hopping_constants
+    site_rates_high = zeros(T, num_sites(spatial_system))
+    bd = (bracket_data === nothing) ? BracketData{T, eltype(starting_state)}() :
+         bracket_data
+    u_low_high = LowHigh(starting_state)
+
+    DirectCRRSSAJumpAggregation(next_jump, next_jump_time, end_time, bd, u_low_high,
+        rx_rates, hop_rates,
+        site_rates_high, save_positions, rng, spatial_system;
+        num_specs = num_species, kwargs...)
+end
+
+# set up a new simulation and calculate the first jump / jump time
+function initialize!(p::DirectCRRSSAJumpAggregation, integrator, u, params, t)
+    p.end_time = integrator.sol.prob.tspan[2]
+    fill_rates_and_get_times!(p, integrator, t)
+    generate_jumps!(p, integrator, u, params, t)
+    nothing
+end
+
+# calculate the next jump / jump time
+function generate_jumps!(p::DirectCRRSSAJumpAggregation, integrator, u, params, t)
+    @unpack rng, rt, site_rates_high, rx_rates, hop_rates, spatial_system = p
+    time_delta = zero(t)
+    while true
+        site = sample(rt, site_rates_high, rng)
+        jump = sample_jump_direct(rx_rates.high, hop_rates.high, site, spatial_system, rng)
+        time_delta += randexp(rng)
+        if accept_jump(p, u, jump)
+            p.next_jump_time = t + time_delta / groupsum(rt)
+            p.next_jump = jump
+            break
+        end
```

Comment:
Does this handle if there is no next jump?

Reply:
To be honest, I do not rembmer any of the logic by now. Do we have a test for what you are asking? If not, we should, for all SSAs.